StyleCop.Analyzers/StyleCop.Analyzers/Settings/SettingsHelper.cs

```diff
@@ -86,7 +134,22 @@ internal static StyleCopSettings GetStyleCopSettings(this AnalyzerOptions option
         /// <returns>A <see cref="StyleCopSettings"/> instance that represents the StyleCop settings for the given context.</returns>
         internal static StyleCopSettings GetStyleCopSettings(this CompilationAnalysisContext context, SyntaxTree tree, DeserializationFailureBehavior failureBehavior, CancellationToken cancellationToken)
         {
-            return GetSettings(context.Options, tree, failureBehavior, cancellationToken);
+            var settingsFile = GetSettingsFile(context.Options, GetJsonValue, cancellationToken);
+            return GetSettings(context.Options, tree, settingsFile, failureBehavior);
+
+            Lazy<JsonValue> GetJsonValue(SourceText settingsText)
```

Comment:
💡 It would be better to make this a `static` lambda in the call to `GetSettingsFile` so it can cache the delegate

Reply:
You lost me :-) This local function captures the context. I hadn't heard of static lambdas before, but it seems the context makes it impossible for it to be static. I am probably misunderstanding you (or static lambdas). Could you elaborate, @sharwell?

Reply:
Suppose you need to capture a local variable, like this:

```cs
int specialValue = 5;
InvokeHelper(() => specialValue);

// helper
int InvokeHelper(Func<int> calculateValue)
  => -calculateValue();
```

You can avoid the delegate allocation by making a generic state argument, and then making the lambda `static`:

```cs
int specialValue = 5;
InvokeHelper(static specialValue => specialValue, specialValue);

// helper
int InvokeHelper<TArg>(Func<TArg, int> calculateValue, TArg arg)
  => -calculateValue(arg);
```

Reply:
Ah, got it! So if I change GetSettingsFile to accept a generic state/context parameter the way you suggest, would there be any difference between a static lambda and making the existing local function static? I just feel that a local function is better for readability.

Reply:
I'm not sure. We'd need to look at the IL output from the compiler to verify.

Reply:
I am not fluent in IL, but it looks like a branch around the allocation if it has already been performed. Similar code for both static cases.
https://sharplab.io/#v2:EYLgtghglgdgNAFxFANnAJiA1AHwAIBMAjALABQeAzAASHUDC1A3udW7TXgCzUBiA9vwAUsBNQAeASmqt2LMu0XUA+tQC81ACLDxcWkQBs1AJ7qAfCckBuWWwC+tmQvZVaPAEIQATiJhipTorySuyqGtpCutQAMtaOivEuhtSiMb5ixtJqFsY2zvaOjq4AKlr8ADzFZkKlAG4QKACuAKZ6eACslXpV1BAAxghQ/DCSjsEhtADsvQNDMEL1Tc1x+dQOZHZAA=

Reply:
Handled in #3648